qiskit/qasm2/__init__.py

```diff
@@ -0,0 +1,523 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2023.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+r"""
+================================
+OpenQASM 2 (:mod:`qiskit.qasm2`)
+================================
+
+Qiskit has support for interoperation with OpenQASM 2.0 programs, both parsing into Qiskit formats
+and exporting back to OpenQASM 2.  The parsing components live in this module, while currently the
+export capabilities are limited to being the :meth:`.QuantumCircuit.qasm` method.
+
+.. note::
+
+    OpenQASM 2 is a simple language, and not suitable for general serialisation of Qiskit objects.
+    See :ref:`some discussion of alternatives below <qasm2-alternatives>`, if that is what you are
+    looking for.
+
+Parsing API
+===========
+
+This module contains two public functions, both of which create a :class:`.QuantumCircuit` from an
+OpenQASM 2 program. :func:`load` takes a filename, while :func:`loads` takes the program itself as a
+string.  Their internals are very similar, so both offer almost the same API.
+
+.. autofunction:: load
+
+.. autofunction:: loads
+
+Both of these loading functions also take an argument ``include_path``, which is an iterable of
+directory names to use when searching for files in ``include`` statements.  The directories are
+tried from index 0 onwards, and the first match is used.  The import ``qelib1.inc`` is treated
+specially; it is always found before looking in the include path, and contains exactly the content
+of the `paper describing the OpenQASM 2 language <https://arxiv.org/abs/1707.03429>`__.  The gates
+in this include file are mapped to standard gates provided by Qiskit.
+
+.. _qasm2-custom-instructions:
+
+Specifying custom instructions
+------------------------------
+
+You can extend the quantum components of the OpenQASM 2 language by passing an iterable of
+information on custom instructions as the argument ``custom_instructions``.  In files that have
+compatible definitions for these instructions, the given ``constructor`` will be used in place of
+whatever other handling :mod:`qiskit.qasm2` would have done.  These instructions may optionally be
+marked as ``builtin``, which causes them to not require an ``opaque`` or ``gate`` declaration, but
```

Comment:
`builtin` is a keyword argument to the function - I made it mono to match the Qiskit style (on the original documentation, it's italic)

Reply:
I left this as-is because the description in the `autoclass` below includes the full signature and explanation of this parameter, and I didn't want to duplicate things in the prose.